### ISSUE TITLE
Restarting nodes in cluster leads to cluster split

### ISSUE THREAD

**Issue by joggeli34** [emqx/emqx#11593]
### What happened?

We have a cluster with three nodes running on kubernetes. When we first start the cluster, all seams to work correctly. 

When we restart each node one after the other with a pause of approximately 2 minutes, the cluster ends in a split state. The nodes can not connect correctly together anymore.

There are some warnings and errors in the log:
```log
2023-09-12T10:07:12.623791+00:00 [warning] 'global' at node 'emqx-dev@emqx-dev-0.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' requested disconnect from node 'emqx-dev@emqx-dev-2.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' in order to prevent overlapping partitions
2023-09-12T10:07:12.623940+00:00 [warning] 'global' at node 'emqx-dev@emqx-dev-0.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' disconnected node 'emqx-dev@emqx-dev-1.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' in order to prevent overlapping partitions
2023-09-12T10:07:18.466718+00:00 [warning] 'global' at 'emqx-dev@emqx-dev-0.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' failed to connect to 'emqx-dev@emqx-dev-2.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local'
2023-09-12T10:07:25.471551+00:00 [error] FORMAT ERROR: "Ekka(AutoCluster): Bad nodes found: ~p, ResL: " - [['emqx-dev@emqx-dev-1.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local'],[{member,'emqx-dev@emqx-dev-0.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local',undefined,<<0,6,5,38,155,25,49,128,143,242,0,0,8,188,0,0>>,1490231354,up,running,{1694,513229,279644},core}]]
2023-09-12T10:07:37.333900+00:00 [warning] msg: Stopping mria, mfa: mria:stop/1, line: 134, reason: join
```

### What did you expect to happen?

Each node should be able to restart without provoking a cluster split. For the connected client, apart from short disconnect, nothing should be visible.

### How can we reproduce it (as minimally and precisely as possible)?

We can reproduce it by running the the 3 node cluster with helm chart. Then we trigger a restart of each node one after the other.

### Anything else we need to know?

The attached logs are from the newly started container. First the node emqx-dev-2 was restarted, then emqx-dev-1 and at last emqx-dev-0

### EMQX version

**5.2.0**, but also happens on 5.1.6

Could not be reproduced on 5.0.21

### OS version

<details>

Helm-chart 5.2.0 running on aks cluster.

</details>

### Log files

[emqx-dev-0.txt](https://github.com/emqx/emqx/files/12585997/emqx-dev-0.txt)
[emqx-dev-1.txt](https://github.com/emqx/emqx/files/12585998/emqx-dev-1.txt)
[emqx-dev-2.txt](https://github.com/emqx/emqx/files/12585999/emqx-dev-2.txt)

**Comment by zmstone** [comment]
Hi @joggeli34 
Thank you for reporting it.
We suspect it has something to do with the clustering network partition detection mechanism.
It's yet to be confirmed in our tests, but could you try to set this config:

`EMQX_NODE__dist_net_ticktime=10s`
This is a hidden config, and the default value is `2m` which happens to be the same as your pause between node restarts.

**Comment by zmstone** [comment]
Can also try to append this line to the end of `/opt/emqx/etc/vm.args`

`-kernel prevent_overlapping_partitions false`

This is a new config for Erlang/OTP 25 which is not yet made configurable from emqx.conf and env variables.
Meaning you'll need to mount the `/opt/emqx/etc/vm.args` file with the above line appended.

EMQX v5.0 runs on Erlang/OTP 24, v5.1 runs on Erlang/OTP 25
In 25, there the network partition detection changed a bit, setting the above config should revert the behavior to 24.

**Comment by ieQu1** [comment]
Sorry, I underestimated the severity of changes in the OTP25 that @zmstone referred to. I am preparing a fix that will revert things to status quo: https://github.com/emqx/emqx/pull/11595

**Comment by ieQu1** [comment]
Can you test the same scenario with the following environment variable added to all the EMQX containers?

`ERL_FLAGS="-kernel prevent_overlapping_partitions false"`

**Comment by zmstone** [comment]
I cannot seem to reproduce it.
I used docker commands directly though.
The node stop (or container kill) always results in an immediate detection of node down, so no race can be found in my test steps.

Suspecting it might be connected to certain networking condition.

**Comment by joggeli34** [comment]
Thanks for the fast answers.

I tried it now with the ```ERL_FLAGS="-kernel prevent_overlapping_partitions false"```. Then I still have some error messages, as somehow the network uses a bit of time until it can connect correctly. But it seems that the split is heals faster. This still leads to some messages losts and a short partial downtime.

As the discovery of the new node is done by dns and the pod changes the ip on start, could that lead to the "old" pod still trying to connect to the old ip instead of the new one for a period of time, which leads to this problem?

For completeness, this is the configuration for the helm chart we are using:

```yaml
# must be ordered-ready so tht minReadySeconds is applied
podManagementPolicy: OrderedReady
# must be ready 2 minutes before the next node gets updated
minReadySeconds: 60
replicaCount: 3
recreatePods: true
metrics:
  enabled: true
pdb:
  enabled: true
resources:
  resources:
    requests:
      cpu: 200m
      memory: 256Mi
    limits:
      cpu: 1
      memory: 1Gi

affinity:
  podAntiAffinity:
    preferredDuringSchedulingIgnoredDuringExecution:
      # schedule in different zones
      - podAffinityTerm:
          labelSelector:
            matchLabels:
              app.kubernetes.io/name: emqx
          topologyKey: topology.kubernetes.io/zone
        weight: 100
      # schedule on different hosts
      - podAffinityTerm:
          labelSelector:
            matchLabels:
              app.kubernetes.io/name: emqx
          topologyKey: kubernetes.io/hostname
        weight: 100

emqxConfig:
  EMQX_DASHBOARD__DEFAULT_PASSWORD: xxxxxxxxxxxxx
  EMQX_NODE__COOKIE: xxxxxxxxxxxxxx

  # Authentication
  authentication.1.mechanism: password_based
  authentication.1.backend: http
  authentication.1.enable: "true"
  authentication.1.method: post
  authentication.1.url: http://{{ .Release.Name }}-authorization/mqtt/authentication
  authentication.1.body: |
    {
      clientid = \"${clientid}\"
      username = \"${username}\"
      password = \"${password}\"
    }

  # Authorization
  authorization.no_match: deny
  authorization.deny_action: disconnect
  authorization.cache.enable: "true"
  authorization.cache.max_size: 128
  authorization.cache.ttl: 1m
  authorization.sources.1.enable: "false"
  authorization.sources.1.type: file
  authorization.sources.1.path: "etc/acl.conf"
  authorization.sources.2.type: http
  authorization.sources.2.enable: true
  authorization.sources.2.method: post
  authorization.sources.2.url: http://{{ .Release.Name }}-authorization/mqtt/authorization
  authorization.sources.2.body: |
    {
      clientid = \"${clientid}\"
      username = \"${username}\"
      topic = \"${topic}\"
      action = \"${action}\"
    }
```

Here the log files from this try:
[emqx-dev-0.txt](https://github.com/emqx/emqx/files/12594950/emqx-dev-0.txt)
[emqx-dev-1.txt](https://github.com/emqx/emqx/files/12594951/emqx-dev-1.txt)
[emqx-dev-2.txt](https://github.com/emqx/emqx/files/12594952/emqx-dev-2.txt)

**Comment by zmstone** [comment]
Hi @joggeli34 
Thank you for testing it out.
We managed to confirm that `ERL_FLAGS="-kernel prevent_overlapping_partitions false"` can help to reduce the risk of cluster-split. We will make it a default config in `v5.2.1` soon.

About slow recovery, I think you have a good point, DNS delay could probably cause late liveness detection for the newly started pod.

**Comment by ieQu1** [comment]
>  But it seems that the split is heals faster.

From the original description of the issue I got an impression that the cluster never healed. Just to clarify: did it eventually heal without `ERL_FLAGS=...`, or?

**Comment by zmstone** [comment]
> > But it seems that the split is heals faster.
> 
> From the original description of the issue I got an impression that the cluster never healed. Just to clarify: did it eventually heal without `ERL_FLAGS=...`, or?

I had the same impression, but after my tests, I believe 5.2.0 should be able to heal, but not 5.1.x.

**Comment by zmstone** [comment]
Inclined to have this issue closed if the `prevent_overlapping_partitions` config helps.
Feel free to open a new one if there are other findings.

**Comment by joggeli34** [comment]
The cluster only sometimes healed without the `ERL_FLAGS`. Probably half the time it healed, half the time it stayed splitted. Once it run for at least 1h without heal.

**Comment by zmstone** [comment]
> The cluster only sometimes healed without the `ERL_FLAGS`. Probably half the time it healed, half the time it stayed splitted. Once it run for at least 1h without heal.

Thank you for the additional info.
Id like to confirm that the unheal situation happen to version 5.2.0?
If it’s 5.1, we can confirm the issue is already identified, if it’s 5.2, we’ll need to take closer look.

**Comment by joggeli34** [comment]
It happend on 5.2, as we directly went from 5.0.21 to 5.2 on one of the cluster where the split happend. I could try to provoque it on our dev cluster with 5.2 tomorrow. Does it help, probably with a lower log level?

**Comment by zmstone** [comment]
Hi. Yes, would be good if you can enable debug level logging.
Thank you.

**Comment by joggeli34** [comment]
Today I tried to reproduce the behavior with the node split. It didn't happen the exact same way as the last times where it didn't heal by it self. But I got the state, where it healed only due to kubernetes terminating one of the pod that didn't have a ready state. Shortly before, the logs looked as following:

[emqx-dev-0.txt](https://github.com/emqx/emqx/files/12621654/emqx-dev-0.txt)
[emqx-dev-1.txt](https://github.com/emqx/emqx/files/12621655/emqx-dev-1.txt)
[emqx-dev-2.txt](https://github.com/emqx/emqx/files/12621656/emqx-dev-2.txt)

They are with notice level, as there is quite some traffic spamming the log when we set it to debug.

**Comment by ieQu1** [comment]
Thanks, we'll take a look at how to improve the behavior.

**Comment by zmstone** [comment]
Sorry for the late follow up.
There are `in order to prevent overlapping partitions` logs in the logs shared on Sep.15 2023.
I suspect that `prevent_overlapping_partitions` might not take effect as expected.

e.g. from [emqx-dev-1.txt](https://github.com/emqx/emqx/files/12621655/emqx-dev-1.txt)
```
2023-09-15T16:02:38.120346+00:00 [warning] 'global' at node 'emqx-dev@emqx-dev-1.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' requested disconnect from node 'emqx-dev@emqx-dev-0.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' in order to prevent overlapping partitions
2023-09-15T16:02:45.186503+00:00 [warning] 'global' at node 'emqx-dev@emqx-dev-1.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' requested disconnect from node 'emqx-dev@emqx-dev-0.emqx-dev-headless.xyz-mqtt-emqx.svc.cluster.local' in order to prevent overlapping partitions
```

After this fix: https://github.com/emqx/emqx/pull/11595 (released in 5.2) this flag is now set by default.

Please feel free to create a new report if the issue is still found in later versions.
Thank you.
